pkg/providers/tinkerbell/controlplane.go

```diff
@@ -0,0 +1,204 @@
+	Secrets []*corev1.Secret
+}
+
+// Objects returns the control plane objects associated with the Tinkerbell cluster.
+func (p ControlPlane) Objects() []kubernetes.Object {
+	o := p.BaseControlPlane.Objects()
+	o = getSecrets(o, p.Secrets)
+
+	return o
+}
+
+func getSecrets(o []kubernetes.Object, secrets []*corev1.Secret) []kubernetes.Object {
+	for _, s := range secrets {
+		o = append(o, s)
+	}
+	return o
+}
+
+func getControlPlaneMachineSpec(clusterSpec *cluster.Spec, diskExtractor *hardware.DiskExtractor) (*anywherev1.TinkerbellMachineConfigSpec, error) {
+	var controlPlaneMachineSpec *anywherev1.TinkerbellMachineConfigSpec
+	if clusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef != nil && clusterSpec.TinkerbellMachineConfigs[clusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name] != nil {
+		controlPlaneMachineSpec = &clusterSpec.TinkerbellMachineConfigs[clusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name].Spec
+		err := diskExtractor.Register(controlPlaneMachineSpec.HardwareSelector)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return controlPlaneMachineSpec, nil
+}
+
+func getWorkerNodeGroupMachineSpec(clusterSpec *cluster.Spec, diskExtractor *hardware.DiskExtractor) (map[string]anywherev1.TinkerbellMachineConfigSpec, error) {
+	var workerNodeGroupMachineSpec *anywherev1.TinkerbellMachineConfigSpec
+	workerNodeGroupMachineSpecs := make(map[string]anywherev1.TinkerbellMachineConfigSpec, len(clusterSpec.TinkerbellMachineConfigs))
+	for _, wnConfig := range clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {
+		if wnConfig.MachineGroupRef != nil && clusterSpec.TinkerbellMachineConfigs[wnConfig.MachineGroupRef.Name] != nil {
+			workerNodeGroupMachineSpec = &clusterSpec.TinkerbellMachineConfigs[wnConfig.MachineGroupRef.Name].Spec
+			workerNodeGroupMachineSpecs[wnConfig.MachineGroupRef.Name] = *workerNodeGroupMachineSpec
+			err := diskExtractor.Register(workerNodeGroupMachineSpecs[wnConfig.MachineGroupRef.Name].HardwareSelector)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return workerNodeGroupMachineSpecs, nil
+}
+
+func getEtcdMachineSpec(clusterSpec *cluster.Spec, diskExtractor *hardware.DiskExtractor) (*anywherev1.TinkerbellMachineConfigSpec, error) {
+	var etcdMachineSpec *anywherev1.TinkerbellMachineConfigSpec
+	if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration != nil {
+		if clusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef != nil && clusterSpec.TinkerbellMachineConfigs[clusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name] != nil {
+			etcdMachineSpec = &clusterSpec.TinkerbellMachineConfigs[clusterSpec.Cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec
+			err := diskExtractor.Register(etcdMachineSpec.HardwareSelector)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return etcdMachineSpec, nil
+}
+
+func generateTemplateBuilder(clusterSpec *cluster.Spec) (providers.TemplateBuilder, error) {
```

Comment:
What do you think about moving all of these methods needed to generate the template builder to `template.go`?